### PR TITLE
Avoid data loss for query items

### LINKED ISSUE #53654: Encoded Plus-Sign in Style URL of vector tiles connection is URL-encoded in wrong way
### What is the bug or the crash?

I have to provide a "Style URL" for  a vector tile connection, which contains a media-type parameter `f=application%2Fvnd.geoserver.mbstyle%2Bjson.` Geoserver e.g. is using those URLs for the OGC Styles API.
If I enter `https://mygeoserver/styles/highways_mb?f=application%2Fvnd.geoserver.mbstyle%2Bjson` into "Style-URL" field, the field-content is immediately decoded to `https://mygeoserver/styles/highways_mb?f=application/vnd.geoserver.mbstyle+json` if closing the dialog. One can see it if reopenig the dialog again. The %2B must not be converted to "+" because the plus-sign is converted to a space by default server-URL-processing (the plus sign is a shorthand notation for a space, see W3C [www.w3.org/Addressing/URL/uri-spec.html](https://www.w3.org/Addressing/URL/uri-spec.html), Query strings )

### Steps to reproduce the issue

Provide a OGC Tiles API over geoserver

QGIS:

- Create a new Vector Tiles Connection
- Provide a valid vector-tiles-URL, e.g. `https://mygeoserver/tiles/collections/highways/tiles/WebMercatorQuad/{z}/{y}/{x}?f=application/vnd.mapbox-vector-tile`
- Provide a Style-URL containing an encoded Plus (%2B), e.g. `https://mygeoserver/styles/highways_mb?f=application%2Fvnd.geoserver.mbstyle%2Bjson`

Even if `https://mygeoserver/styles/highways_mb?f=application%2Fvnd.geoserver.mbstyle%2Bjson` is a valid Url for getting the style, the server delivers an error because `https://mygeoserver/styles/highways_mb?f=application/vnd.geoserver.mbstyle json` isn't a valid Url.

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.30.3-'s-Hertogenbosch | QGIS code revision | 2411cec1
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.5
GDAL/OGR version | 3.7.0
PROJ version | 9.2.0
EPSG Registry database version | v10.082 (2023-02-06)
GEOS version | 3.11.2-CAPI-1.17.2
SQLite version | 3.41.1
PDAL version | 2.5.3
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.13.1
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins
excel_sync | 3.6
infas_lt_geocoder | 1.0
pluginbuilder3 | 3.2.1
plugin_reloader | 0.9.1
qgis2web | 3.16.0
topojson_writer | 0.1.1
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.6
processing | 2.12.99

</body></html>QGIS version
3.30.3-'s-Hertogenbosch
QGIS code revision
[2411cec1](https://github.com/qgis/QGIS/commit/2411cec1)
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.7.0
PROJ version
9.2.0
EPSG Registry database version
v10.082 (2023-02-06)
GEOS version
3.11.2-CAPI-1.17.2
SQLite version
3.41.1
PDAL version
2.5.3
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins
excel_sync
3.6
infas_lt_geocoder
1.0
pluginbuilder3
3.2.1
plugin_reloader
0.9.1
qgis2web
3.16.0
topojson_writer
0.1.1
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [ ] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by pathmapper**
Fix https://github.com/qgis/QGIS/issues/53654

Use the [default](https://doc.qt.io/qt-6/qurlquery.html#queryItems) `PrettyDecoded` instead of `FullyDecoded` for query items.

From the [docs](https://doc.qt.io/qt-5/qurl.html#query):

> Note that use of [QUrl::FullyDecoded](https://doc.qt.io/qt-5/qurl.html#ComponentFormattingOption-enum) in queries is discouraged, as queries often contain data that is supposed to remain percent-encoded, **including the use of the "%2B" sequence to represent a plus character ('+')**.

Test added.

### CODE REVIEW

File: tests/src/core/testqgsdatasourceuri.cpp [hunk 1]
```diff
@@ -492,6 +492,11 @@ void TestQgsDataSourceUri::checkAuthParams()
   QCOMPARE( uri4.param( QStringLiteral( "password" ) ), QStringLiteral( "😁😂😍" ) );
   QCOMPARE( uri4.password(), QStringLiteral( "😁😂😍" ) );
 
+  // issue GH #53654
+  QgsDataSourceUri uri5;
+  uri5.setEncodedUri( QStringLiteral( "zmax=14&zmin=0&styleUrl=http://localhost:8000/&f=application%2Fvnd.geoserver.mbstyle%2Bjson" ) );
+  QCOMPARE( uri5.param( QStringLiteral( "f" ) ), QStringLiteral( "application%2Fvnd.geoserver.mbstyle%2Bjson" ) );
+
```
Comment (nyalldawson) [inline tests/src/core/testqgsdatasourceuri.cpp:499]: Could use also add a test where an explicit "+" character is used in the value for a parameter?

Comment (pathmapper) [reply tests/src/core/testqgsdatasourceuri.cpp:499]: Good idea, done in https://github.com/qgis/QGIS/pull/53701/commits/4bb4979a1021509c63d63d3acb3464e17f609157.

Comment (nyalldawson) [reply tests/src/core/testqgsdatasourceuri.cpp:499]: Sorry - one last test: can you create a QgsDataSourceUri, and then set two parameters for it using setParam, one with a %2B literal in the value and another with a + literal, and see if the values survive the round trip through encodedUri/setEncodedUri losslessly?

Comment (pathmapper) [reply tests/src/core/testqgsdatasourceuri.cpp:499]: :+1: , they do survive, test added in https://github.com/qgis/QGIS/pull/53701/commits/8aa299c569acebb5a4c62f2792f672ff57544d79.

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Thank you!

**Pull request merged**